### ISSUE TITLE
Come up with a recommended way to render/store tutorials

### ISSUE THREAD

**Issue by grst** [scverse/cookiecutter-scverse#40]
I think that wherever it is computationally feasible, tutorials should be run as a CI check and the tutorial use the version rendered by the CI to avoid failing or outdated tutorials. 

With the current template it is already possible to have tutorials built by `nbshpinx` on `readthedocs.org`, but it would be nice to have something that scales better. Also it needs to be documented somewhere how to enable this behavior. 

- [ ] Recommendations where to store tutorials (submodule for executed `ipynb` files when not using CI builds?)
- [ ] Documentation how to enable CI builds
- [ ] Ideally, each notebook is built in a separate worker (-> parallelization) (-> individual checkmark for each notebook instead of digging into sphinx logs) 
- [ ] GPU support

See also #19.

**Comment by grst** [comment]
I'd want something like

```mermaid
flowchart TD
   A[push to PR] --> B[run tests]
   A --> C[build tutorials on GH actions]
   C --> D[trigger RTD build]
```

**Comment by grst** [comment]
@michalk8, can you elaborate how the squidpy setup works? And maybe point me to a PR where the tutorials were actually built, I only find PRs where that particular action was skipped.

**Comment by grst** [comment]
Inspired by the pre-commit CI, the following approach could work: 

1. trigger CI in PR by push
2. rerun notebooks. If the output changes compared to the previous version, add a commit with the rendered ipynb files (either in main repo or submodule)
3. the automatic commit triggers RTD again. Other CI checks can be skipped.

**Comment by grst** [comment]
> Inspired by the pre-commit CI, the following approach could work:
> 
> 1. trigger CI in PR by push
> 2. rerun notebooks. If the output changes compared to the previous version, add a commit with the rendered ipynb files (either in main repo or submodule)
> 3. the automatic commit triggers RTD again. Other CI checks can be skipped.

I think I'll give this approach a try in one of my repos.

**Comment by grst** [comment]
[MyST-nb](https://myst-nb.readthedocs.io/en/latest/quickstart.html) might be a good alternative to nbsphinx. 

Once benefit is that it [supports caching](https://myst-nb.readthedocs.io/en/latest/computation/execute.html) through [jupyter-cache](https://github.com/executablebooks/jupyter-cache). It could be nice to restore that cache when running notebooks through github actions.

**Comment by flying-sheep** [comment]
Isaac wants submodules back in scanpy, see https://github.com/scverse/scanpy/issues/2636

I think we should plan the best approach here first (or approach*es*: Might be that different projects have different needs).

> store the notebooks in the main repo in text format

Probably not. Bespoke markdown dialects are a bad bad thing. Y’all like MyST, I’m not going to go back on it, some “I told you so”s nonwithstanding. But I’ll need some *excellent* arguments to be convinced that storing notebooks in any other format than .ipynb might be a good idea. I’m very happy that I don’t do R anymore, since Rmd instead of .ipynb cost the whole ecosystem so much potential synergy.

**Comment by grst** [comment]
> I think we should plan the best approach here first (or approaches: Might be that different projects have different needs).

agreed. Might need a temporary solution in scanpy that could be reverting to the submodule until we have something everyone is happy with. 

> No, unless you have some excellent arguments for it that you haven’t shared yet

The main argument for me is that
 * text format are easy to review on github 
 * it is easy to fix merge conflict if there are any. 

I think there are two formats that are particularly interesting and have excellent editor support. I also don't think jupytext with >6k stars is a niche tool. 

1. plain python script ([python percent format])(https://jupytext.readthedocs.io/en/latest/formats-scripts.html#the-percent-format))

Out-of-the-box support in Spyder, vscode, pycharm, vim+slime. It's just a python script, i.e. all python tooling works on it (no additional tooling required to apply ruff, black etc)

2. Quarto markdown

The evolution of Rmd, but language agnostic. Rstudio/posit fully supports python with it. Native support in rstudio and vscode (via official extension). 

But it's not the hill I am going do die on, I can live with an ipynb with stripped out outputs. The main selling point is that tutorials without output (in whatever format) can be kept in the same repo as the main documentation.

**Comment by flying-sheep** [comment]
> Might need a temporary solution in scanpy that could be reverting to the submodule until we have something everyone is happy with.

As said there, if someone wants that, they can do it, I don’t think it has a significant enough advantage to spend effort on it personally 😄 

> text format are easy to review on github

Good argument. There’s multiple extensions to do it ([e.g.](https://gitnotebooks.com/)), but of course they aren’t seamlessly integrated in GitHub. The question is how much that matters.

> it is easy to fix merge conflict if there are any.

I never had problems with this, nbdime is great at this. Granted, you need to always do it locally instead of being able to click the GitHub UI button most of the time.

> [python percent format](https://jupytext.readthedocs.io/en/latest/formats-scripts.html#the-percent-format)

- No markdown preview while authoring, confusing layout while authoring. That’s not great, since you can accidentally break the magic comment syntax and the corruption can be overlooked while reviewing the notebook. Standard problems when you pretend that text is a data structure. (That’s why I don’t use bash/zsh anymore)
- No standalone 3rd party tooling without converting them to .ipynb first. I assume that’s not great if you want to use some standard tooling to modify/lint/whatever the file. Maybe I’m wrong and things can be seamlessly converted back and forth without loss of information.

  The jupterlab plugin makes this a bit less painful, but it’s kinda dumb that .py files get the Jupyter icon and get opened as notebooks.

> Quarto markdown

Still one language-specific editor’s proprietary format that hasn’t garnered tool support outside of the R world. Nope.

I predicted then that Rmd would become big enough in the R world that it would prevent the scientific landscape to grow together and create interoperability with tools that work on the same format. It made me sad. I was right.

**Comment by grst** [comment]
Happy to go with stripped ipynb if you prefer. Technically it would be easy to support both at some point, as it's probably just a jupytext call in step 1.

**Comment by grst** [comment]
@flying-sheep, maybe an even easier solution than writing a sphinx plugin is to use [DVC](https://dvc.org/). It's basically "git for data", similar to git lfs, but with a bunch of supported backends (e.g. S3 bucket). I tried it at work for some data analysis projects and it's pretty neat. 

Basically
 * you can `dvc add` any file in the repository. It will then be gitignored and insteada `*.dvc` file is checked into git that contains the file hash with which it can be retreived from the storage backend
 * there's a `dvc repro` command that executes commands to "reproduce the analysis" that are stored in a yaml file. This could be used to render notebooks on the CI. 

And on readthedocs, one would just add `dvc pull` as a pre-build command, and it would retreive the rendered `.ipynb` files from wherever they are stored.

**Comment by flying-sheep** [comment]
I’m not a fan of that approach either.

I think they should just be moved back into the scanpy repo. With `git clone --filter=blob:none ...`, nobody needs to download big amounts of data, and it’s less work to do that once than wrangle the submodule.

@ivirshup you said it was decided to go that route and seemed unwilling to go back on that decision. Can you link to where that’s discussed? I’d like to see what submodules bring to the table.

**Comment by grst** [comment]
uv apparently does not yet use `--filter=blob:none`, but `pip` does, see 
https://github.com/scverse/scverse-tutorials/issues/60#issuecomment-2730928667

So sooner or later it will be there. The best route forwards seems therefore to be to put the tutorials in the main repo and implementing a CI to build them as described above. That means no sphinx-plugin or other hacks necessary. Basically

1. commit notebooks (with or without outputs) to repo
2. CI executes all changed notebooks. If output changed, CI makes another commit

That's basically it. The tricky part seems access tokens again. Using a github token doesn't work on forks easily. There's two possible approaches: 
  * use a two-step approach (executing notebooks without token, upload artifact, commit in second stage on `pull-request-target` (https://securitylab.github.com/resources/github-actions-preventing-pwn-requests/)
  * use a github app

But the github app, we'd kind of have to offer centrally? Also not sure if we would then have to pay for EC2 instances or if we could still use github actions runners?

**Comment by flying-sheep** [comment]
@LucaMarconato just said in the core developer meeting that he’d like to look into this, I’m happy to help.
